src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -840,6 +856,57 @@ impl Tree {
         node.children.is_empty() && !live.contains(&node.tag)
     }
 
+    /// Checks whether a node can be replaced by its only child.
+    /// If so, returns the index of said only child.
+    /// If not, returns none.
+    fn can_be_replaced_by_single_child(
+        &self,
+        idx: UniIndex,
+        live: &FxHashSet<BorTag>,
+    ) -> Option<UniIndex> {
+        let node = self.nodes.get(idx).unwrap();
+        if node.children.len() != 1 || live.contains(&node.tag) || node.parent.is_none() {
+            return None;
+        }
+        // Since protected nodes are never GC'd (see `borrow_tracker::GlobalStateInner::visit_provenance`),
+        // we know that `node` is not protected because otherwise `live` would
+        // have contained `node.tag`.
```

Comment:
Can you add a debug assertion checking that the node is not protected? Or do we not have access to that state?

Reply:
That requires passing extra state around, unfortunately.

Reply:
Okay, in that case please add a comment in `FrameExtra::visit_provenance` saying that visiting the protected nodes here is load-bearing for the TB GC. Also this comment here should reference `FrameExtra::visit_provenance`, not `GlobalStateInner`.